dbt/macros/get_s3_dependency_dir.sql

```diff
@@ -0,0 +1,48 @@
+-- Get the S3 location that is used to store Python model dependencies
+-- for the current target
+{% macro get_s3_dependency_dir() %}
+    {{
+        return(
+            _get_s3_dependency_dir(
+                target,
+                env_var,
+                exceptions.raise_compiler_error,
+            )
+        )
+    }}
+{% endmacro %}
+
+{% macro _get_s3_dependency_dir(target, env_var_func, raise_error_func) %}
+    {% set dir_suffix = "" %}
+    {% if target.name == "dev" %}
+        {% set username = env_var_func("USER") %}
+        {% if not username %}
+            {{
+                return(
+                    raise_error_func("USER env var must be set when target is 'dev'")
+                )
+            }}
+        {% endif %}
+        {% set dir_suffix = "/" ~ username %}
+    {% elif target.name == "ci" %}
+        {% set head_ref = slugify(env_var_func("HEAD_REF")) %}
+        {% if not head_ref %}
+            {{
+                return(
+                    raise_error_func(
+                        "HEAD_REF env var must be set when target is 'ci'"
+                    )
+                )
+            }}
+        {% endif %}
+        {% set dir_suffix = "/" ~ head_ref %}
+    {% endif %}
+    {% set s3_dependency_dir = target.s3_data_dir ~ "packages" ~ dir_suffix %}
```

Comment:
One downside of this approach is that it stores packages _inside_ the data dir, when ideally they would be stored alongside the data dir. For clarity, here are the three paths for the three different targets we support:

* `dev`
  * data dir:  `s3://ccao-dbt-athena-dev-us-east-1/data/`
  * dependency dir: `s3://ccao-dbt-athena-dev-us-east-1/data/packages/<username>/`
* `ci`
  * data dir: `s3://ccao-dbt-athena-ci-us-east-1/data/`
  * dependency dir: `s3://ccao-dbt-athena-ci-us-east-1/data/packages/<branch_slug>/`
* `prod`
  * data dir: `s3://ccao-athena-ctas-us-east-1/`
  * dependency dir: `s3://ccao-athena-ctas-us-east-1/packages/`

I think `prod` is probably the most reasonable of the three. They all feel acceptable right now, but not perfect. What do you think?

Reply:
**suggestion (non-blocking):** Well, if we're following our existing schema for the dev, ci, prod split, something like this might be better:

* `dev`
  * data dir:  `s3://ccao-dbt-athena-dev-us-east-1/data/`
  * dependency dir: `s3://ccao-dbt-athena-dev-us-east-1/packages/<username>/`
* `ci`
  * data dir: `s3://ccao-dbt-athena-ci-us-east-1/data/`
  * dependency dir: `s3://ccao-dbt-athena-ci-us-east-1/packages/<branch_slug>/`
* `prod`
  * data dir: `s3://ccao-athena-ctas-us-east-1/`
  * dependency dir: `s3://ccao-athena-packages-us-east-1/`

Reply:
@dfsnow I agree, that pattern would be cleaner, but [per my comment above](https://github.com/ccao-data/data-architecture/pull/435/files#r1596033934) we aren't actually able to configure the dependency directory as a dedicated profile variable, so it needs to be derivable from other variables like `s3_data_dir`. This is why I proposed the organization sketched out in this PR, as imperfect as it is; if you have suggestions for a better one that is derivable from existing profile variables I would be grateful!

Reply:
I like that idea! Done in 02cc83a3ee89b3bda571f489d9f04a125ab069b3.